### ISSUE TITLE
PDF with Password Encryption using plsql

### ISSUE THREAD

**Issue by bebikashg** [Pilooz/pl_fpdf#13]
Is there any way to create a password projected pdf file using PL/SQL.

**Comment by Maxwbh** [comment]
Yes, it is possible entirely in PL/SQL, without Java and without an external
service.

PDF password protection is the standard security handler from the spec: an
`/Encrypt` dictionary in the trailer, a key derived from the user and owner
passwords plus the file ID and the permission bits, and then every string and
every stream in the document encrypted with that key. Nothing in it needs a
library — it is arithmetic and byte handling, which is why it can be written in
PL/SQL.

Two things are worth knowing before you build or adopt anything:

**RC4 is not a reasonable choice any more.** It has been broken for years and
**PDF 2.0 removed it from the specification**; current readers warn about it or
refuse it outright. If you are starting now, use AES — AES-128 (`/AESV2`) or
AES-256 (`/AESV3`, revision 6).

**A password is not a signature.** Encryption controls who can open the file. It
does not tell you whether the content was altered, and it does not prove who
produced it. If what you need is tamper evidence or authorship, you need a
digital signature, which is a different problem entirely. Also note the owner
password only restricts what a *cooperating* reader allows — printing, copying —
and any reader is free to ignore it. The user password is the part that actually
protects the file.

The code in this repository does not implement this. A fork at
[`Maxwbh/pl_fpdf`](https://github.com/Maxwbh/pl_fpdf) added it in its 3.x line,
in case it saves you the work:

```sql
l_encrypted := PL_FPDF.EncryptPDF(
  p_pdf            => l_pdf,
  p_user_password  => 'user123',
  p_owner_password => 'owner456',
  p_permissions    => JSON_OBJECT_T('{"print":true,"copy":false}'),
  p_encryption     => 'AES-256'
);
```

It supports RC4-40, RC4-128, AES-128 and AES-256, with user and owner passwords
and eight permission flags, and there is a matching `DecryptPDF`, plus
`IsEncrypted` and `GetSecurityInfo` for inspecting a file you did not produce.

Two implementation details that may matter to you: it does **not** depend on
`DBMS_CRYPTO`, which is often not granted in a plain schema — the hashes come
from `STANDARD_HASH` and the ciphers are implemented in the package itself. And
a PDF 1.5+ source, with cross-reference streams and object streams, is flattened
on encryption: objects inside object streams become top-level objects and the
output carries a classic cross-reference table.

If you would rather implement it yourself, the parts to get right are the key
derivation (algorithms 2 through 7 of the spec, and 2.A/2.B for revision 6), the
per-object key, and remembering that strings inside dictionaries are encrypted
too — not just streams. Validating the output against an independent reader
rather than against your own decryptor is what separates "the file says it is
protected" from "the file is protected".
